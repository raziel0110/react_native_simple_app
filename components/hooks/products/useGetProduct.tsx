import {useState, useEffect} from 'react';
import axios from 'axios';

const useGetProduct = (url: string) => {
  const [data, setData] = useState({});

  useEffect(() => {
    const fetchData = async () => {
      if (url) {
        const response = await axios.get(url);
        setData(response.data);
      }
    };

    fetchData();
  }, [url]);

  return data;
};

export default useGetProduct;
