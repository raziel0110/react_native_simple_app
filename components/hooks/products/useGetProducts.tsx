import {useQuery} from 'react-query';
import axios from 'axios';

const PRODUCTS_URL = 'https://dummyjson.com/products';

const getProducts = async (url: string) => {
  return await axios.get(url);
};

export const useGetProducts = (searchKey: string) => {
  const url =
    searchKey.length > 0
      ? `${PRODUCTS_URL}/search?q=${searchKey}`
      : PRODUCTS_URL;
  return useQuery(['products', url], () => getProducts(url));
};
