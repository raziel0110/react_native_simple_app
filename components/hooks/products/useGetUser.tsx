import axios from "axios";
import { useEffect, useState } from "react";

const PROFILE_URL = 'https://dummyjson.com/auth/me'

export interface DataUI {
  image?: string;
  company?: {
    title?: string;
    department?: string;
  };
  lastName?: string;
  firstName?: string;
  email?: string;
  phone?: string;
  username?: string;
}

export const useGetUser = (token: string) => {  
  const [data, setData] = useState({});
  console.log("TOKEN =>", token)
  useEffect(() => {
    const fetchData = async () => {
      try {
        const response = await axios.get(PROFILE_URL, {headers: {'Authorization': `Bearer ${token}`}})
        console.log(response)
        setData(response.data)
      } catch (err: any) {
        console.log("err =>",err)
        setData({isError: true, error: err.response})
      }
    }
    fetchData();
  }, [token])

  return data
}