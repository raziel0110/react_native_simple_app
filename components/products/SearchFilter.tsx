import React from 'react';
import {StyleSheet, TextInput, View} from 'react-native';
import FontAwesomeIcon from 'react-native-vector-icons/FontAwesome';

const SearchFilter = ({onChangeSearchKey}: any) => {
  return (
    <View style={style.searchContainer}>
      <TextInput
        placeholder="Search Product"
        style={style.inputContainer}
        onChangeText={val => onChangeSearchKey(val)}
      />
      <FontAwesomeIcon
        name="search"
        size={20}
        color="lightgray"
        style={style.searchIcon}
      />
    </View>
  );
};

const style = StyleSheet.create({
  searchContainer: {
    position: 'relative',
  },
  searchIcon: {
    position: 'absolute',
    top: 0,
    left: 0,
  },
  inputContainer: {
    marginLeft: 25,
    fontSize: 16,
    color: 'gray',
  },
});

export default SearchFilter;
