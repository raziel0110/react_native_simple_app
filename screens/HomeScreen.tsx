import React, {useState} from 'react';
import {
  View,
  Text,
  SafeAreaView,
  FlatList,
  TouchableOpacity,
  StyleSheet,
  ActivityIndicator,
} from 'react-native';
import {
  ViewContainer,
  ImageContainer,
  DescriptionView,
  BottomDescription,
  PriceBlock,
  TitleHeader,
} from './HomeScreenStyle';
import FontAwesomeIcon from 'react-native-vector-icons/FontAwesome';
import SearchFilter from '../components/products/SearchFilter';
import {useGetProducts} from '../components/hooks/products/useGetProducts';
import useDebounce from '../components/hooks/products/useDebounce';

const BASE_URL = 'https://dummyjson.com/products';

const HomeScreen = (props: {
  navigation: {navigate: (arg0: string, arg1: {itemId: any}) => void};
}): React.JSX.Element => {
  const [searchKey, setSearchKey] = useState('');
  const debounced: string = useDebounce(searchKey, 500);
  const url = searchKey.length > 0 ? `${BASE_URL}/search?q=${debounced}`
  : `${BASE_URL}`;

  const {isLoading, data, hasNextPage, fetchNextPage, isFetching} = useGetProducts(url);
  const keyExtractor = (_: any, index: number) => index.toString();
  const dataArr = data?.pages?.map(page => page.data.products).flat();

  const showItem = (item: any) => {
    props.navigation.navigate('Item', {
      itemId: item.id,
    });
  };

  const emptyList = () => {
    if (dataArr?.length === 0) {
      return (
        <View style={{display: 'flex', flexDirection: 'row', alignItems: 'center', justifyContent: 'center', height:200, width:300, marginTop: 'auto', marginBottom: 'auto', marginLeft: 'auto', marginRight: 'auto'}}>
          <FontAwesomeIcon name="exclamation-circle" size={30} style={{color: 'red'}}/>
          <Text style={{marginLeft: 10, fontSize: 16, fontWeight: 'bold'}}>No Items Found!!!</Text>
        </View>
      )
    }

    return null;
  }

  const renderLoader = () => {
    if (isFetching) {
      return (
        <View style={{marginBottom: 20, marginTop: 20, padding: 10}}>
          <ActivityIndicator size="large"/>
        </View>
      )
    }
    return null;
  }

  const renderItem = ({item}: any) => (
    <TouchableOpacity onPress={() => showItem(item)}>
      <ViewContainer>
        <View>
          <ImageContainer source={{uri: item.thumbnail}} resizeMode="cover" />
        </View>
        <View>
          <TitleHeader>{item.title.substring(0,30)}</TitleHeader>
          <DescriptionView>
            <View>
              <BottomDescription>Stock: {item.stock}</BottomDescription>
            </View>
            <View>
              <Text>
                Price:<PriceBlock>{item.price}$</PriceBlock>
              </Text>
            </View>
          </DescriptionView>
        </View>
      </ViewContainer>
    </TouchableOpacity>
  );

  return (
    <SafeAreaView>
      <View>
        <View style={styles.filterContainer}>
          <SearchFilter onChangeSearchKey={setSearchKey} />
        </View>
        <FlatList
          data={dataArr}
          keyExtractor={keyExtractor}
          renderItem={renderItem}
          onEndReachedThreshold={0}
          onEndReached={() => {
            if (hasNextPage && !isLoading) {
              fetchNextPage()
            }
          }}
          ListEmptyComponent={emptyList}
          ListFooterComponent={renderLoader}
        />
      </View>
    </SafeAreaView>
  );
};

const styles = StyleSheet.create({
  filterContainer: {
    marginTop: 5,
    marginBottom: 7,
    marginLeft: 10,
    padding: 10,
    backgroundColor: '#fff',
    width: '95%',
    borderRadius: 10,
  },
  spinner: {
    flex: 1,
    justifyContent: 'center',
  },
});

export default HomeScreen;
