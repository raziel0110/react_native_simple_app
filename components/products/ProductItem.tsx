import React from 'react';
import { StyleSheet, Text, View, Image, Animated, TouchableOpacity, Button } from 'react-native';
import { BottomDescription, PriceBlock } from '../../screens/HomeScreenStyle';
import { Dimensions } from 'react-native';
import { Swipeable, TextInput } from 'react-native-gesture-handler';
import FontAwesomeIcon from 'react-native-vector-icons/FontAwesome';
import { useDispatch } from 'react-redux';
import { removeItemCart, addQuantity, removeQuantity } from '../../context/features/checkoutSlice';


const width = Dimensions.get('screen').width;

interface CheckoutCart {
  thumbnail: string;
  title: string;
  stock: number;
  price: number;
  quantity: number;
}

const styles = StyleSheet.create({
  imageContainer: {
    width: 75,
    height: 75,
    borderRadius: 15,
    marginRight: 24,
    shadowColor: '#294B29',
    shadowOpacity: 0.3,
  },
  titleHeader: {
    fontSize: 18,
    fontWeight: 'bold',
  },
  itemContainer: {
    padding: 15,
    display: 'flex',
    width: width, 
    flexDirection: 'row',
    marginBottom: 10,
    backgroundColor: 'white',
    shadowOpacity: 0.2,
    shadowOffset: {width: 1, height: 7},
    shadowColor: '#294B29',
  },
  descriptionContainer: {
    marginTop: 8,
    display: 'flex',
    justifyContent: 'space-between',
    flexDirection: 'row',
  },
  quantityButton: {
    marginLeft: 10,
    marginRight: 10,
    height: 20,
    width: 30,
    backgroundColor: '#294B29',
    display: 'flex', 
    justifyContent: 'center',
    alignItems: 'center',
    borderRadius: 5
  },
  quantityText: {
    color: 'white',
    fontWeight: '700',
  }
});

const ProductItem = (props: {item: CheckoutCart}): React.JSX.Element => {
  const {item} = props;
  const dispatch = useDispatch();

  const onDeleteItem = () => {
    dispatch(removeItemCart(item));
  }

  const onRemoveQuantity = () => {
    dispatch(removeQuantity(item));
  }

  const onAddQuantity = () => {
    dispatch(addQuantity(item));
  }

  const rightSwipe = (_progress: string | number, dragX: { interpolate: (arg0: { inputRange: number[]; outputRange: number[]; extrapolate: string; }) => any; }) => {
    const scale = dragX.interpolate({
      inputRange: [0, 100],
      outputRange: [0, 1],
      extrapolate: 'clamp'
    })
    return (
      <TouchableOpacity onPress={onDeleteItem}>
        <View style={{backgroundColor: 'rgba(255,99,125, 0.7)', height: 105, width: 60, display: 'flex', alignItems: 'center', justifyContent: 'center'}}>
          <Animated.Text style={{transform: [{scale: scale}]}}>
            <FontAwesomeIcon name="trash" size={30} color="white" />
          </Animated.Text>
        </View>
      </TouchableOpacity>
    )
  }

  return (
    <Swipeable renderLeftActions={rightSwipe}>
      <View style={styles.itemContainer}>
        <View>
          <Image source={{uri: item.thumbnail}} resizeMode="cover" style={styles.imageContainer}/>
        </View>
        <View>
          <Text style={styles.titleHeader}>{item.title}</Text>
          <View style={styles.descriptionContainer}>
            <View>
              <BottomDescription>Stock: {item.stock}</BottomDescription>
            </View>
            <View>
              <Text>
                Price:<PriceBlock>{item.price}$</PriceBlock>
              </Text>
            </View>
          </View>
          <View style={{marginTop: 5}}>
            <View style={{display: 'flex', flexDirection: 'row'}}>
              <Text>Quantity: </Text>
              <View style={{display: 'flex', flexDirection: 'row'}}>
                <TouchableOpacity onPress={onRemoveQuantity} style={styles.quantityButton}>
                  <Text style={styles.quantityText}>-</Text>
                </TouchableOpacity>
                <Text>{item.quantity}</Text>
                <TouchableOpacity onPress={onAddQuantity} style={styles.quantityButton}>
                  <Text style={styles.quantityText}>+</Text>
                </TouchableOpacity>
              </View>
            </View>
          </View>
        </View>
      </View>
    </Swipeable>
  )
};

export default ProductItem;