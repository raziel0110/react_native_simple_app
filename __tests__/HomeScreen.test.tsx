import 'react-native';
import React from 'react';
import HomeScreen from '../screens/HomeScreen';
import { NavigationContainer } from '@react-navigation/native';
import { render, screen, waitFor } from '@testing-library/react-native';
import { QueryClient, QueryClientProvider } from 'react-query';
import * as ProductsApi from '../components/hooks/products/useGetProducts';

const queryClient = new QueryClient();
queryClient.setDefaultOptions({ queries: { cacheTime: 0 } })

const wrapper = ({ children }: any) => (
    <QueryClientProvider client={queryClient}>{children}</QueryClientProvider>
);

describe("<HomeScreen />", () => {
  // const getProductsMocked = jest.spyOn(ProductsApi, 'useGetProducts')
  const products = [
    {
      id: 1,
      thumbnail: 'http://test.com/thumbnail1.jpg',
      title: 'test 1',
      stock: 12,
      price: 10
    },
    {
      id: 2,
      thumbnail: 'http://test.com/thumbnail2.jpg',
      title: 'test 2',
      stock: 12,
      price: 10
    }
  ];
  const mock = {isLoading: false, isSuccess: true, data: {pages: [{data: products}]}, isFetching: false}
  const mockedService = jest.mock('react-query', () => ({
    useInfiniteQuery: jest.fn().mockReturnValue((mock))
  }))

  beforeEach( async () => {

    
    await waitFor(() => {
      render(
        <NavigationContainer>
            <HomeScreen navigation={{ navigate: jest.fn() }} />
        </NavigationContainer>, {wrapper}
      );
    });
  });
  afterEach(() => { queryClient.clear() })
  
  it('render correctly', () => {
    expect(screen.getByPlaceholderText('Search Product')).toBeTruthy();
  })

  // it('call the get product hook', () => {
  //   expect(mockedService).toHaveBeenCalled();
  // });

  // it('fetch products in pages', () => {
  //   // products.forEach(prod => {
  //   //   expect(screen.getByText(prod['title'])).toBe(prod['title']);
  //   // })
  // })

  test("ViewAllProducts", () => {
    render(
      <NavigationContainer>
          <HomeScreen navigation={{ navigate: jest.fn() }} />
      </NavigationContainer>, {wrapper}
    );
    screen.debug();
    expect(screen.toJSON()).toMatchSnapshot()
  })
})
