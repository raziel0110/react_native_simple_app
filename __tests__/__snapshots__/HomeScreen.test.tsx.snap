// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<HomeScreen /> ViewAllProducts 1`] = `
<RCTSafeAreaView>
  <View
    accessibilityIgnoresInvertColors={true}
  >
    <Image
      blurRadius={35}
      resizeMode="cover"
      source={
        {
          "testUri": "../../../assets/background.jpg",
        }
      }
      style={
        [
          {
            "bottom": 0,
            "left": 0,
            "position": "absolute",
            "right": 0,
            "top": 0,
          },
          {
            "height": undefined,
            "width": undefined,
          },
          undefined,
        ]
      }
    />
    <View>
      <View
        id="search-filter"
        style={
          {
            "backgroundColor": "#fff",
            "borderRadius": 10,
            "marginBottom": 7,
            "marginLeft": 10,
            "marginTop": 5,
            "padding": 10,
            "width": "95%",
          }
        }
      >
        <View
          style={
            {
              "position": "relative",
            }
          }
        >
          <TextInput
            onChangeText={[Function]}
            placeholder="Search Product"
            style={
              {
                "color": "gray",
                "fontSize": 16,
                "marginLeft": 25,
              }
            }
          />
          <Text
            allowFontScaling={false}
            selectable={false}
            style={
              [
                {
                  "color": "lightgray",
                  "fontSize": 20,
                },
                {
                  "left": 0,
                  "position": "absolute",
                  "top": 0,
                },
                {
                  "fontFamily": "FontAwesome",
                  "fontStyle": "normal",
                  "fontWeight": "normal",
                },
                {},
              ]
            }
          >
            ï€‚
          </Text>
        </View>
      </View>
      <RCTScrollView
        ListEmptyComponent={[Function]}
        ListFooterComponent={[Function]}
        getItem={[Function]}
        getItemCount={[Function]}
        keyExtractor={[Function]}
        onContentSizeChange={[Function]}
        onEndReached={[Function]}
        onEndReachedThreshold={0}
        onLayout={[Function]}
        onMomentumScrollBegin={[Function]}
        onMomentumScrollEnd={[Function]}
        onScroll={[Function]}
        onScrollBeginDrag={[Function]}
        onScrollEndDrag={[Function]}
        removeClippedSubviews={false}
        renderItem={[Function]}
        scrollEventThrottle={0.0001}
        stickyHeaderIndices={[]}
        viewabilityConfigCallbackPairs={[]}
      >
        <View>
          <View
            onLayout={[Function]}
          >
            <View
              style={
                {
                  "marginBottom": 20,
                  "marginTop": 20,
                  "padding": 10,
                }
              }
            >
              <ActivityIndicator
                size="large"
              />
            </View>
          </View>
        </View>
      </RCTScrollView>
    </View>
  </View>
</RCTSafeAreaView>
`;
